lib/puppet/util/execution.rb

```diff
@@ -319,6 +322,18 @@ class << self
   # @api private
   #
   def self.execute_posix(command, options, stdin, stdout, stderr)
+    exec_options = {}
+    exec_options[:chdir] = options[:cwd] if options[:cwd]
+
+    # NOTE: The reason we do not put this check in 'execute' is because
```

Comment:
Yeah, that's something I wanted to check with @Iristyle on because on my machine, it does do the right thing (UNC paths and regular '\' paths both work).

Reply:
Yeah, I think we want to use `Puppet::FileSystem.directory?` here rather than Ruby file system calls as @branan notes. Always default to using the `Puppet::FileSystem` calls because they fix problems in the stock Ruby implementations of equivalent functionality.

Reply:
UPDATE: 

Looking closer at the FileSystem code, it's a little tricky to understand what's going on as it looks like we're using the default Ruby implementation behind our `directory?` helper. I think that may actually be flawed as I'm quite certain it won't work properly with reparse points (at least on older Ruby versions - I haven't checked 2.5). That's something we might want to change, orthogonal to this PR.

Interestingly we have a helper called `dir_exist?` as well. That call also looks like it might use default behavior, but it doesn't quite. It funnels to the Windows specific `exist?` implementation at https://github.com/puppetlabs/puppet/blob/master/lib/puppet/file_system/windows.rb#L34

That code calls in to https://github.com/puppetlabs/puppet/blob/master/lib/puppet/util/windows/file.rb#L115 which *is* reparse point aware, which is what we want in this case to be the most tolerant of Windows directory alternatives (that act like directories)

Reply:
Also note that since this code path is `execute_posix` this discussion is a bit moot... but it might make sense to move the check up higher in the call stack so that it impacts both Windows and POSIX.